### ISSUE TITLE
Compare branch-tip with branch-base (not branch-tip^)

### ISSUE THREAD

**Issue by thomie** [nomeata/gipeda#21]
The overview page for branches, such as the one at https://perf.haskell.org/ghc/, should compare branch tips with branch bases. It currently compares the branch tips with their parent commit, which is not the right thing when a branch contains multiple commits not on master.

For example, the `wip/pretty` branch is currently colored green, but closer inspection shows it should be colored dark red. The `wip/pretty` branch contains 2 commits not present on master.

First [commit](https://perf.haskell.org/ghc/#revision/cb828c0d43148dfdfa407a82310279ce98601233) on `wip/pretty`:

```
tests/alloc/T1969   575205848   + 11.01%    638535056   bytes
tests/alloc/T3064   241185432   + 3.73% 250182344   bytes
tests/alloc/T3294   2661675144  + 25.90%    3351169320  bytes
tests/alloc/T4801   395203440   + 15.36%    455919080   bytes
tests/alloc/T5321FD     427386568   + 6.08% 453358104   bytes
tests/alloc/T5321Fun    449961544   + 5.78% 475962416   bytes
tests/alloc/T5837   35517072    + 13.86%    40438320    bytes
tests/alloc/T783    468652880   + 11.75%    523717152   bytes
tests/alloc/T9961   680274680   + 3.16% 701745240   bytes
```

Second [commit](https://perf.haskell.org/ghc/#revision/8e62060a28f041b267a004967ad8efa3cb465500) on `wip/pretty` (this is currently shown in green on the front page):

```
tests/alloc/T1969   638535056   - 5.91% 600808984   bytes
tests/alloc/T3294   3351169320  - 13.44%    2900740384  bytes
tests/alloc/T4801   455919080   - 9.00% 414885752   bytes
tests/alloc/T5321FD     453358104   - 3.60% 437038064   bytes
tests/alloc/T5321Fun    475962416   - 3.41% 459714496   bytes
tests/alloc/T783    523717152   - 6.67% 488804472   bytes
```

Both [commits](https://perf.haskell.org/ghc/#compare/a5061a96724922097e4181d452a64618e35fa297/8e62060a28f041b267a004967ad8efa3cb465500) on `wip/pretty` (this should be shown in red on the front page instead):

```
tests/alloc/T1969   575935048   + 4.32% 600808984   bytes
tests/alloc/T3294   2662804168  + 8.94% 2900740384  bytes
tests/alloc/T4801   396623880   + 4.6%  414885752   bytes
tests/alloc/T5837   35621192    + 13.52%    40438032    bytes
tests/alloc/T783    470239352   + 3.95% 488804472   bytes
```

I have the following git alias in my .gitconfig, from http://stackoverflow.com/a/4991675, maybe it can be used for the implementation:

```
oldest-ancestor = !bash -c 'diff --old-line-format= --new-line-format= <(git rev-list --first-parent \"${1:-master}\") <(git rev-list --first-parent \"${2:-HEAD }\") | head -1' -
```

**Comment by nomeata** [comment]
So in essence, the clicking on a branch in the main view will not go to the per-branch view, but rather directly to the comparison view with a suitable base commit? (which is not always easy to identify, I worry).

**Comment by thomie** [comment]
Yes, clicking on a branch should go to the comparison view with a suitable base commit.

See the stackoverflow link I posted for a way to find that commit.

**Comment by nomeata** [comment]
Eugh, that is some ugly code there. I’d really prefer something well defined in terms of git’s own tools. `git merge-base` doesn’t cut it?

**Comment by thomie** [comment]
I started writing a comment here trying to explain why `merge-base` indeed doesn't cut it, but I now think it's actually better than the command I posted above, for our purposes.

Given the following git commit history:

```
-- X -- A -- B -- C -- D ("master" branch) 
          \         \     
           \         \   
             G -- H -- I -- J ("topic" branch)
```

`git merge-base master topic` will find commit `C`. The command I posted above goes through a lot of trouble to find commit `A`.

In this case comparing commit `J` with commit `C` is exactly what we want, effectively measuring the combined performance effect of commits `G`, `H`, `I` and `J`, but not including commits `B` and `C`, which were merged in from master (and which the command posted above would include).

I am assuming the merge here didn't have any conflicts, and we never merge topic branches (partially) back into master.

In the more common case without any merges whatsoever, depicted below, both solutions will find commit `A`, and `git merge-base` does so much faster:

```
-- X -- A -- B -- C -- D ("master" branch) 
          \              
           \            
             G -- H -- I -- J ("topic" branch)
```
